### ISSUE TITLE
the ezLCDlib breaks SotwareSerial and probably other things

### ISSUE THREAD

**Issue by Jassper** [earthmake/arLCD#1]
I spent two days trying to get my stuff to work only to find out that making a single call to your lib is breaking SoftwareSerial. Here is a test example;
[code]
# include <ezLCDLib.h>
# include<SoftwareSerial.h>

const byte numChars = 96;
char receivedChars[numChars];   // an array to store the received data

boolean newData = false;

SoftwareSerial SSerial(4,5);
ezLCD3 lcd;

void setup()
{
        SSerial.begin(9600);
        Serial.begin(9600);
    Serial.println("<Arduino is ready>");
        lcd.cls(BLACK);
}

void loop()
{
    recvWithEndMarker();
        digitalWrite(13,LOW);
    showNewData();
}

void recvWithEndMarker()
{
    digitalWrite(13,HIGH);
        static byte ndx = 0;
    char endMarker = '\n';
    char rc;

```
if (SSerial.available() > 0)
            {
    rc = SSerial.read();

    if (rc != endMarker) 
                    {
        receivedChars[ndx] = rc;
        ndx++;
        if (ndx >= numChars) 
                            {
            ndx = numChars - 1;
                }
              }
    else 
                    {
        receivedChars[ndx] = '\0'; // terminate the string
        ndx = 0;
        newData = true;
            }
        }
```

}

void showNewData()
{
    if (newData == true) 
        {
        Serial.print("This just in ... ");
        Serial.println(receivedChars);
        newData = false;
    }
}
[/code]

This freezes the program and nothing works after the lcd.cls(BLACK) command. Comment that out and everything works again. What happens is the LED on pin 13 comes on, but never goes out and no data is ever reported on the hardware serial. This has GOT to be fixed, becasue I am useing the Hardware serial to talk to the LCD I am left with no choice but to use the SoftwareSerial to communicate to the world. 

Thanks

**Comment by Dev49net** [comment]
I don't think this library is breaking SoftwareSerial. I think it is "breaking" the regular Serial, because the LCD works through Serial connection. So you cannot have two Serial connections on the same ports - one with computer and one with LCD.
I started working with it today and I also quickly found out that this LCD cannot be used with regular Serial. Bot SoftwareSerial should be working well, I see no reason why this lib may interfere with it.

P.S. @earthmake: Why do you keep the library zip file on GitHub? On GitHub you should keep unzipped files, it is a code sharing service, not file hosting...

**Comment by Jassper** [comment]
I know I can't have two serial connections on the same port - that is why I am using SoftwareSerial witch allows me to add a second port (not physically) using pins 4 and 5 (see above).
Software  Serial here, http://arduino.cc/en/Reference/softwareSerial

To get it to work I had to increase the serial buffer form 64 to 128 in the SoftwareSerial.h file. I think because the LCD lib has sooooo many serial.print() commands that it really slows down the processor and the Software Serial gets missed.

Here is the full history of my problem
http://forum.arduino.cc/index.php?topic=296333.0

I don't know why Earthlink doesn't use the I2C interface and leave the hardware serial port open for the user. Instead they rob us of the one serial port the Uno and Pro mini have.

**Comment by Dev49net** [comment]
Well, we have to keep in mind that LCDs are not the strongest points of Arduino. There just isn't enough power in them. And whether the communication happens through Serial or I2C, doesn't change much because after connecting the LCD, your Arduino will be barely able to do anything more. If you have a complicated project that not only uses LCD but also has some complicated algorithms in the backend, plus it's communicating with other devices through SoftwareSerial, it definitely means that Arduino is not the right tool for this job. Even if it works, you would be able to do it on some stronger platform much more comfortably (BeagleBone maybe?).

> To get it to work I had to increase the serial buffer form 64 to 128 in the SoftwareSerial.h file. I think because the LCD lib has sooooo many serial.print() commands that it really slows down the processor and the Software Serial gets missed.

I don't think that's how it works. ArLCD has its own GPU and the serial data sent to it is very limited. It's just basic commands, I don't think Arduino sends to the GPU much more messages through Serial than the commands we are writing using their supplied library. It's the GPU that does all the math there, while Arduino is free to do other stuff. At least these are my observation after using it for 2 days. I am building a MIDI data monitor with LCD and it responds to MIDI data sent on SoftwareSerial pretty well so far. The LCD is there to display list of messages and also sliders with current MIDI values and I'm pleasantly surprised with the performance.
Maybe you are sending data to your LCD in the loop? If that's the case, then yes, you are constantly sending data through Serial.

EDIT: Have you tried to use I2C for your communication, instead of SoftwareSerial? Maybe you would be able to get better results with it.

**Comment by q2dg** [comment]
Well, it seems pretty abandoned...
